File: src/main/java/de/lukaseichler/recurrentneuralnetwork/LayerResults.java
package de.lukaseichler.recurrentneuralnetwork;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import com.google.common.base.Preconditions;
import com.google.common.collect.Iterables;
import com.google.common.collect.ListMultimap;
import com.google.common.collect.MultimapBuilder;
import de.lukaseichler.recurrentneuralnetwork.structure.Layer;

/**
 * @author leichler
 * TODO improving performance by implementing a custom datastructure
 */
public class LayerResults {

    private final ListMultimap<Layer, Double> resultsPerLayer = MultimapBuilder
            .linkedHashKeys()
            .arrayListValues()
            .build();

    public int size() {
        return resultsPerLayer.keys().size();
    }

    public void add(@Nullable final Layer layer, @Nonnull final List<Double> results) {
        Preconditions.checkNotNull(results);
        resultsPerLayer.putAll(layer, results);
    }

    public Collection<Double> get(@Nonnull final Layer layer) {
        Preconditions.checkNotNull(layer);
        return resultsPerLayer.get(layer);
    }

    public @Nullable List<Double> getLastResult() {
        if (resultsPerLayer.keySet().size() == 0) {
            return null;
        }
        return resultsPerLayer.get(Iterables.getLast(resultsPerLayer.keySet()));
    }

    public @Nullable List<Double> getSecondLastResult() {
        if (resultsPerLayer.keySet().size() < 2) {
            return null;
        }
        return resultsPerLayer.get(Iterables.get(resultsPerLayer.keySet(), resultsPerLayer.keySet().size() - 2, null));
    }

    @SuppressWarnings("unchecked") public Iterator<List<Double>> getReverseValueIterator() {
        LinkedList<List<Double>> values = resultsPerLayer.keySet().stream().map(resultsPerLayer::get).collect(Collectors.toCollection(LinkedList::new));
        return values.descendingIterator();
    }

    public void clear() {
        resultsPerLayer.clear();
    }


}

File: src/main/scala/de/lukaseichler/recurrentneuralnetwork/LayerResults.scala
package de.lukaseichler.recurrentneuralnetwork

import javax.annotation.{Nonnull, Nullable}

import de.lukaseichler.recurrentneuralnetwork.structure.Layer

import scala.collection.mutable

/**
 * @author Lukas Eichler

 */
class LayerResults {
  private val resultsPerLayer = new mutable.LinkedHashMap[Layer, mutable.Set[Double]] with mutable.MultiMap[Layer, Double]

  def size: Int = {
    resultsPerLayer.keys.size
  }

  def add(@Nullable layer: Layer, @Nonnull results: List[Double]) {
    require(results != null)
    results.foreach(x => resultsPerLayer.addBinding(layer, x))
  }

  def get(@Nullable layer: Layer): Set[Double] = {
    resultsPerLayer.apply(layer).toSet
  }

  @Nullable def getLastResult: List[Double] = {
    if (resultsPerLayer.keySet.size == 0) {
      return null
    }
    resultsPerLayer.apply(resultsPerLayer.keySet.last).toList
  }

  @Nullable def getSecondLastResult: Set[Double] = {
    if (resultsPerLayer.keySet.size < 2) {
      return null
    }
    resultsPerLayer.apply(resultsPerLayer.keySet.toList.reverse(1)).toSet
  }

  @SuppressWarnings(Array("unchecked")) def getReverseValueIterator: Iterator[List[Double]] = {
    val values: List[List[Double]] = resultsPerLayer.keySet.map(x => resultsPerLayer.apply(x).toList).toList
    values.reverseIterator
  }

  def clear() {
    resultsPerLayer.clear()
  }
}
